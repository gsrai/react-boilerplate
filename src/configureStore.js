import { compose, createStore, applyMiddleware } from 'redux'
import createSagaMiddleware, { END } from 'redux-saga'
import rootReducer from './reducers'
import rootSaga from './sagas'

import { composeWithDevTools } from 'redux-devtools-extension'

export default function configureStore (history, initialState) {
  const sagaMiddleware = createSagaMiddleware()
  let store = null

  if (process.env.NODE_ENV === 'production') {
    store = createStore(
      rootReducer,
      initialState,
      compose(
        applyMiddleware(sagaMiddleware)
      )
    )
  } else {
    const DevToolsExt = typeof (window) !== 'undefined' && window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__() // eslint-disable-line

    store = createStore(
      rootReducer,
      initialState,
      // used composeWithDevTools instead of redux's compose
      // in order to have access to the redux-devtools-extension
      composeWithDevTools(
        applyMiddleware(sagaMiddleware)
      )
    )

    // Hot reload reducers (requires Webpack or Browserify HMR to be enabled)
    if (module.hot) {
      module.hot.accept('./reducers', () => {
        const nextRootReducer = require('./reducers').default // eslint-disable-line global-require
        store.replaceReducer(nextRootReducer)
      })
    }
  }

  store.close = () => store.dispatch(END)
  sagaMiddleware.run(rootSaga)
  return store
}
