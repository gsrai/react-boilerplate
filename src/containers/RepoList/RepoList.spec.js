import { shallow, mount } from 'enzyme'
import React from 'react'

import configureStore from 'redux-mock-store'
import ConnectedComponent, { RepoList } from './RepoList'

describe('Component Rendering', () => {
  const mockStore = configureStore()

  // beforeEach(() => {

  // })

  test('component should render connected component', () => {
    const initialState = createInitialState({})

    const store = mockStore(initialState)
    const wrapper = shallow(<ConnectedComponent store={store} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('component should render default', () => {
    const initialState = createInitialState({})
    const actions = {
      actions: {
        loadRepoListInit: jest.fn(),
        loadRepoList: jest.fn()
      }
    }
    const props = {
      ...initialState.repoListReducer,
      ...actions
    }
    const wrapper = shallow(<RepoList {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('component should dispatch init and load actions on mount', () => {
    const initmock = jest.fn()
    const loadmock = jest.fn()
    const initialState = createInitialState({})
    const actions = {
      actions: {
        loadRepoListInit: initmock,
        loadRepoList: loadmock
      }
    }
    const props = {
      ...initialState.repoListReducer,
      ...actions
    }
    mount(<RepoList {...props} />)
    expect(initmock.mock.calls.length).toBe(1)
    expect(loadmock.mock.calls.length).toBe(1)
    expect(loadmock.mock.calls[0][0]).toEqual('react')
  })

  test('component should render loading text when loading', () => {
    const initialState = createInitialState({
      loading: true
    })

    const actions = {
      actions: {
        loadRepoListInit: jest.fn(),
        loadRepoList: jest.fn()
      }
    }

    const props = {
      ...initialState.repoListReducer,
      ...actions
    }

    const wrapper = shallow(<RepoList {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('component should render repo data when successfully loaded', () => {
    const repoData = [
      {
        full_name: 'facebook/react',
        html_url: 'https://github.com/facebook/react',
        description: 'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
        stargazers_count: 108963
      }
    ]

    const initialState = createInitialState({
      loadSuccess: true,
      repos: repoData
    })

    const actions = {
      actions: {
        loadRepoListInit: jest.fn(),
        loadRepoList: jest.fn()
      }
    }

    const props = {
      ...initialState.repoListReducer,
      ...actions
    }

    const wrapper = shallow(<RepoList {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('component should pass reason when there is an error', () => {
    const reason = 'cannot load repos'
    const initialState = createInitialState({
      loadError: true,
      reason
    })

    const actions = {
      actions: {
        loadRepoListInit: jest.fn(),
        loadRepoList: jest.fn()
      }
    }

    const props = {
      ...initialState.repoListReducer,
      ...actions
    }

    const wrapper = shallow(<RepoList {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})

const createInitialState = (state) => {
  const repoListReducerState = {
    loading: state.loading || false,
    loadSuccess: state.loadSuccess || false,
    loadError: state.loadError || false,
    repos: state.repos || [],
    reason: state.reason || ''
  }

  const initialState = {
    repoListReducer: repoListReducerState
  }

  return initialState
}
