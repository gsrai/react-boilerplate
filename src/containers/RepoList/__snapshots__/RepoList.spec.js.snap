// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Rendering component should pass reason when there is an error 1`] = `
<div>
  <h1
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    Example
  </h1>
  <div>
    A problem has occurred: cannot load repos
  </div>
</div>
`;

exports[`Component Rendering component should render connected component 1`] = `
<RepoList
  actions={
    Object {
      "loadRepoList": [Function],
      "loadRepoListFailure": [Function],
      "loadRepoListInit": [Function],
      "loadRepoListSuccess": [Function],
    }
  }
  loadError={false}
  loadSuccess={false}
  loading={false}
  reason=""
  repos={Array []}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;

exports[`Component Rendering component should render default 1`] = `
<div>
  <h1
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    Example
  </h1>
  <div />
</div>
`;

exports[`Component Rendering component should render loading text when loading 1`] = `
<div>
  <h1
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    Example
  </h1>
  <div>
    loading repositories...
  </div>
</div>
`;

exports[`Component Rendering component should render repo data when successfully loaded 1`] = `
<div>
  <h1
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    Example
  </h1>
  <div>
    <RepoListItem
      key="0"
      repo={
        Object {
          "description": "A declarative, efficient, and flexible JavaScript library for building user interfaces.",
          "full_name": "facebook/react",
          "html_url": "https://github.com/facebook/react",
          "stargazers_count": 108963,
        }
      }
    />
  </div>
</div>
`;
